### DISCUSSION TITLE
Proposal: argument name in Func and Action

### DISCUSSION THREAD

**Discussion by hez2010** [dotnet/csharplang#2738] [Language Ideas]
# Proposal: argument name in Func and Action
Currently when we write `Func<T1, T2, ...> func = (p1, p2, ...) => ...`, we get a delegate without parameter names.  
Therefore, when we want to invoke the delegate, IDE cannot prompt us the parameter name.

```csharp
Func<int, int, int> myFunc = (x, y) => x + y; 
```
When we invoke it:
```
myFunc(  // IDE will prompt us 'arg1'
myFunc(1,  // IDE will prompt us 'arg2'
```

## Solution
Allow argument name in Func and Action, just like what `ValueTuple<T>` did.
```csharp
Func<int num1, int num2, int> myFunc = (x, y) => x + y; 
```

Then when we invoke it:
```
myFunc(  // IDE will prompt us 'num1'
myFunc(1,  // IDE will prompt us 'num2'
```

The same for `Action<T, ...>`.

**Comment by YairHalberstadt** [comment]
I have often wanted to name the types in Dictionaries. For example: `Dictionary<int id, string name>`

I think it wouldn't necessarily be a bad idea to be able to name type arguments in order to provide features for use of the IDE only.

**Comment by svick** [comment]
Related (and more ambitious) proposals: https://github.com/dotnet/csharplang/issues/470, https://github.com/dotnet/csharplang/issues/269.

**Comment by TonyValenti** [comment]
This reminds me of named members on tuples.  Seems like a good idea.

**Comment by HaloFour** [comment]
I'd made the comment on other proposals but this only works in the narrow case where the named generic type argument is only used once for any given member or any given method signature.  Delegates like `Action<...>` and `Func<...>` fit these constraints but only by convention.  Same with tuples.  But what happens when a type doesn't follow that convention?

```cs
public class Foo<T> {
    public void M(T x, T y) { }
}


var f = new Foo<int bar>();
f.M(...) // ?
```

**Comment by YairHalberstadt** [comment]
@HaloFour

If this was a purely IDE thing, I don't think it matters to much. The IDE could choose to display bar for both, bar for neither. Whatever.

The language just changes to allow you to put the name there. The meaning is left to the IDE to decide.

**Comment by HaloFour** [comment]
@YairHalberstadt 

That feels a little weird if the language itself doesn't offer any support beyond just naming the generic type arguments.  It makes the following feel really confusing:

```cs
Func<int x, int y, int y> f = (x, y) => x + y;
int result = f(arg1: 2, arg2: 3);
```

**Comment by YairHalberstadt** [comment]
@HaloFour
Do you mean the fact that the IDE shows them as named arguments, but you can't actually use named arguments?

**Comment by HaloFour** [comment]
@YairHalberstadt  Exactly.

**Comment by magol** [comment]
Related to #894

**Comment by ufcpp** [comment]
Also related to https://github.com/dotnet/csharplang/issues/883

**Comment by ahdung** [comment]
On the contrary, I even want to remove the parameter name of delegate:
`public delegate void Foo(string, int)`

And the damned default value, it cause reflection doesn't respect method's default value.
